### PR TITLE
[core][distributed] support layer size undividable by pp size in pipeline parallel inference

### LINKED ISSUE #6114: [Feature]: support layer size undividable by pp size in pipeline parallel inference
### 🚀 The feature, motivation and pitch

tracking the issue here

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by youkaichao**
fixes https://github.com/vllm-project/vllm/issues/6114

### CODE REVIEW

File: vllm/config.py [hunk 1]
```diff
@@ -276,12 +274,6 @@ def verify_with_parallel_config(
                 "Pipeline parallelism is only supported for the following "
                 f" architectures: {_PP_SUPPORTED_MODELS}.")
 
-        if total_num_hidden_layers % pipeline_parallel_size != 0:
```
Comment (andoorve) [inline vllm/config.py:279]: Shouldn't we error out when `total_num_hidden_layers < pipeline_parallel_size `?

**Review by andoorve — COMMENTED** [review]
There's a subtle point here:

https://github.com/vllm-project/vllm/blob/966fe72141e8365721840b7ababfb78601c23ead/vllm/executor/distributed_gpu_executor.py#L38-L46

We take min of blocks across all workers - as a result of this the GPU mem utilization of 0...n-2th will be slightly lower than expected.

I don't think it's a big deal, just something to be aware of.

File: vllm/config.py (hunk 1 above)
Comment (youkaichao) [reply vllm/config.py:279]: speculative decoding has `total_num_hidden_layers=0` for some statistical worker (e.g. ngram).

File: vllm/worker/worker.py [hunk 2]
```diff
@@ -50,6 +50,7 @@ def __init__(
     ) -> None:
         self.model_config = model_config
         self.parallel_config = parallel_config
+        self.parallel_config.rank = rank
```
Comment (andoorve) [inline vllm/worker/worker.py:53]: How does this get used by the way?

Comment (youkaichao) [reply vllm/worker/worker.py:53]: it is used in `get_num_layers` .

**Comment by youkaichao** [comment]
> We take min of blocks across all workers - as a result of this the GPU mem utilization of 0...n-2th will be slightly lower than expected.

Good point. I also noticed this. Some GPU memory waste, but it is better than not being supported.

**Review by andoorve — APPROVED** [review]
LGTM

**Pull request merged**